Adds a post on what is TypeScript (#552)
artsy/artsy.github.io: _posts/2019-04-05-omakase-typescript.markdown

date: "2019-05-05"
author: [orta]
categories: [community, javascript, roads and bridges, typescript]
---

TypeScript is a language which builds on JavaScript. This post is a non-technical overview of what JavaScript is,
how TypeScript extends it and why we choose to adopt it at Artsy.

<!-- more -->

## What is JavaScript?

First up, you can't describe TypeScript without talking about JavaScript. To create a website (and a bunch of other
types of things) you work in three languages: HTML, CSS and JavaScript (JS). The first two languages about
presenting the look and feel of the website, and JavaScript allows you to edit what it shows and make it
interactive.

We describe having these sets of skills as being a "front-end" developer. You have to understand those three
languages to present anything inside a web browser like Safari, Firefox or Chrome. So, given how popular the web
is, there is a massive demand for people who are good at using these three languages.

There is also the set of skills for the "back-end" developers, which are to create computer servers which
communicate either to a web browser (by passing it HTML/CSS/JS) or to another computer server (by sending a kind of

> JavaScript.) You don't need to use HTML, CSS or JS to write this type of code, but it's usually an end-product of

Author: Aye, I dig there are other methods, but I feel like I want to cram JS as JSON into this somehow (in part to hint that JS is used in a lot of contexts)

> JavaScript.) You don't need to use HTML, CSS or JS to write this type of code, but it's usually an end-product of

Author: > by sending a simpler form of JavaScript

> JavaScript.) You don't need to use HTML, CSS or JS to write this type of code, but it's usually an end-product of

Reviewer: What about "by sending raw data"? It keeps it more general, and there _are_ tonnes of XML APIs out there still 😉

your work. We mostly build our back-ends in Ruby or JavaScript at Artsy.

### What do Programming Languages do?

Programming languages are an interesting problem to solve. People read code many, many multiples of times more than

> they write it - so developers create languages which are good at solving particular problems in a terse amount of

Author: Hrm, I do want to use "terse" here over "small" because that comes with extra context of doing the same thing but with less. Will figure out how people use the word properly

> they write it - so developers create languages which are good at solving particular problems in a terse amount of

Author: small works fine in this context 👍

code. Here's an example using JavaScript:

```js
var name = "Danger"
console.log("Hello, " + danger)
```

The first line makes a variable (something you can store something in) and then the seconds outputs text out to the
console (think DOS, or the terminal) `"Hello, Danger"`. JavaScript is designed to work as a scripting language,
which means the code starts at the top of the file and then goes through line by line. To provide some contrast,
here is the [same behavior](https://repl.it/repls/VioletredGlisteningInfo) in Java, which is built with different
language constraints:

```java
class Main {
  public static void main(String[] args) {
    String name = "Danger";
    System.out.println("Hello, " + name);
  }
}
```

> Note: if you find the naming of Java and JavaScript confusing (it is) - that's because when JavaScript as being
> created, Java was looking to be really the next hot language to work with (it did turn out that way for a few

> > decades, but now JavaScript is usually the first language people have heard of.)

Author: haha, yah, maybe a decade or two

Aside from having a lot more lines, the Java version comes with a lot of words that aren't necessarily about
telling the computer exactly what to do, e.g. `class Main {`, `public static void main(String[] args) {`, `}` and
`}` again. It also has semi-colons at the end of some line. Java is aimed at building different things, and these
extras to the code sense within the constraints of building a Java app.

To get to my main point though, there is one standout line I'd like us to compare:

```
// JavaScript
var name = "Danger"
// Java
String name = "Danger";
```

> Both of these lines set a variable to a be "Danger" as a string (a collection of characters, which in programming

Author: Yes!

we call a string (this [5m video](https://www.youtube.com/watch?v=czTWbdwbt7E) is a good primer)) - in JS you write
`var` which tells the next word after that is the name of the variable. However in Java, the first thing you write
is `String` instead.

There are many values which can go into a variable, and `var` accepts all of them in JS. In Java though, we have to
say upfront what can be inside that variable, which is only strings, so:

```js
// Before in JS
var name = "Danger"
// Also OK
var name = 1
var name = false
var name = { first: "Danger", last: "McMaslen" }

// Before in Java
String name = "Danger";
// Not OK, the code wouldn't be accepted by Java
String name = 1;

// This example isn't quite what you'd write, but it shows the difference well
Map<String, String> person = new HashMap<String, String>();
person.put("first", "Danger");
person.put("last", "McMaslen");

> String name = person;

Reviewer 2: not sure what's going on here. Might need more comments. Also hash maps and generics are probably going to just confuse people who don't know what a type system is.

Author: I really couldn't find some other code that correlates to `var name = { first: "Danger", last: "McMaslen" }` in a reasonable way in Java. But only having two cases (int/bool) felt a bit too few

Author: and I don't think you can skip the generics, but I'll see

Author: I can do an array instead, from the above paragraph

```

These trade-offs make sense in the context for which these languages were built. JavaScript was **originally**
designed in a a short time, to specifically do a small amount of things. Java on the other hand was built
specifically to make very, very large amounts of easy to read with - which means sometimes you have to do more

> writing/reading chores. It's an implicit vs explicit thing.

Author: perfect

### What is TypeScript?

TypeScript is a programming language - it contains all of JavaScript, and then a bit more. So in our case above,
lets compare the scripts for "Hello, Danger" in JavaScript vs TypeScript:

```js
// JavaScript
var name = "Danger"
console.log("Hello, " + danger)

// TypeScript
var name = "Danger"
console.log("Hello, " + danger)

> // Yep, you're not missing something, there's no difference

Author: I  put it in here so that people didn't spend ages trying to figure it before reading on, I know I've done that before :D

```

Due to TypeScript aiming to only _extend_ JavaScript, your normal JavaScript code should work fine with TypeScript.
The parts of the JavaScript language which TypeScript extends are specifically about trying to make it easier to be
more explicit in your code.

```diff
- var name = "Danger"
+ var name: string = "Danger"
console.log("Hello, " + danger)
```

This allows the reader to be certain that `name` will only be a string. Annotating your variables also allows
TypeScript to check this for you.

Simply speaking, we call these annotation "Types" and this is the "Type" in TypeScript. The tag-line for TypeScript
is "JavaScript which scales" which is a statement that these extra type annotations allows you to work on bigger
projects. This is because you can verify how correct your code is, and as many people read and write it then you

> don't need to fit it all in your head.

Reviewer 2: > This is because you can verify how correct your code is, and as many people read and write it then you
don't need to fit it all in your head.

This could be more impactful. You could talk about how much effort it takes to manually find type errors, and how error prone that process is. How TS makes it much faster to orient yourself in a codebase. How it helps to make vast sweeping refactors way less scary...

> TypeScript extends JavaScript more, but this is the core concept - and that's enough for now.

Reviewer 2: I'd just delete this line. It's not helpful.

### Why does Artsy use TypeScript?

Artsy definitely isn't the size of Microsoft! Artsy is about 30 engineers, and Microsoft are about 60,000. However,

> some of our problems are the same. The "front end" developers at Artsy build apps which are thousands of files big.

Reviewer 2: ```suggestion
some of our problems are the same. Developers at Artsy build apps which are made up of thousands of files.
```

> Any single change to one file, could affect any other file. A good way to think about this is how in a story,

Reviewer 2: ```suggestion
Any single change to one file could affect any other file. A good way to think about this is how in a story,
```

> changing a conversation earlier can trigger a butterfly effect of changes later down the line.

Reviewer 2: FWIW I think you could find a better analogy here. Maybe 'ripples on a pond'.

> A change to one individual file can affect the behaviour of any number of other files, like throwing a pebble into a pond and causing ripples to spread out to the bank.

Author: Yep!

> Typically, the need to ensure all your changes do what you want is less of a problem for people building websites.

Reviewer 2: > ensure all your changes do what you want

This might be easier to grok as

> ensure there are no bugs 

though not semantically identical

Author: Yeah, I think bugs is common enough that peeps  will get what I'm going for

Websites are easy to make changes to, because if you change the site - everyone gets the update instantly. However,
we also build our iOS app with JavaScript, and a change to the app requires Apple to review the changes and for
someone to download the new version from the App Store.

This means that the iOS team needs to have more checks that everything is OK before shipping the app to the world.
Using TypeScript gives our team the ability to feel good that the changes we have made are only the changes we
want.

TypeScript isn't the only programming language to tackle the problem of making JavaScript code safer, but it's the
one with the biggest community, allows people to re-use their JavaScript knowledge and has a really good

> programmers text editor.

Reviewer 2: This sentence should be split up. Might be more impactful to talk about 'extra tools to help developers work faster' than 'text editor' specifically.

There are many values which can go into a variable, and `var` accepts all of them in JS. In Java though, we have to
say upfront what can be inside that variable, which is only strings, so:

```js
// Before in JS
var name = "Danger"
// Also OK
var name = 1
var name = false
var name = ["2018-02-03", "2019-01-12"]

// Before in Java
String name = "Danger";
// Not OK, the code wouldn't be accepted by Java
String name = 1;
String name = false
String name = {"2018-02-03", "2019-01-12"};
```

These trade-offs make sense in the context for which these languages were built back in 1995. JavaScript was
originally designed to be a small programming language which handled simple interactions on websites. Java on the
other hand was built specifically to make big apps which could run on any computer. Their needs had different
scales, so the language required different types of code.

Java required programmers to be more explicit with the values of their variables because the programs they expected
people to build were more complex.

### What is TypeScript?

TypeScript is a programming language - it contains all of JavaScript, and then a bit more. Using our example above,

> lets compare the scripts for "Hello, Danger" in JavaScript vs TypeScript:

Reviewer 3: let's

Reviewer 4: ```suggestion
let's compare the scripts for "Hello, Danger" in JavaScript vs TypeScript:
```

```

Due to TypeScript aiming to only _extend_ JavaScript, your normal JavaScript code should work fine with TypeScript.
The things TypeScript adds to JavaScript are intended to help you be more explicit about what kinds of data are
used in your code, a bit like Java.

```diff
- var name = "Danger"
+ var name: string = "Danger"
console.log("Hello, " + danger)
```

This extra `: string` allows the reader to be certain that `name` will only be a string. Annotating your variables
also allows TypeScript to check this for you. This is very useful because keeping track of changes like the type of
value in a variable seems easy when it's one or two, but once it starts hitting the hundreds, that's a lot to keep
track of.

Simply speaking, we call these annotations "Types". Hence the name <i>Type</i>Script. The tag-line for TypeScript
is "JavaScript which scales" which is a statement that these extra type annotations allows you to work on bigger
projects. This is because you can verify up-front how correct your code is. This means you have less need to
understand how every change affects the rest of the program.

### Why does Artsy use TypeScript?

> Artsy definitely isn't the size of Microsoft! Artsy is about 30 engineers, and Microsoft are about 60,000. However,

Reviewer 3: This is the first time you mention Microsoft. Say something before that about why! Where's TS developed ...

JavaScript.) You don't need to use HTML, CSS or JS to write this type of code, but it's usually an end-product of
your work. We mostly build our back-ends in Ruby or JavaScript at Artsy.

### What do Programming Languages do?

Programming languages are an interesting problem to solve. People read code many, many multiples of times more than
they write it - so developers create languages which are good at solving particular problems with a small amount of
code. Here's an example using JavaScript:

```js
var name = "Danger"
console.log("Hello, " + danger)
```

The first line makes a variable (a kind of box you can keep things in) and then the second line outputs text to the
console (think DOS, or the terminal) `"Hello, Danger"`. JavaScript is designed to work as a scripting language,
which means the code starts at the top of the file and then goes through line by line. To provide some contrast,
here is the [same behavior](https://repl.it/repls/VioletredGlisteningInfo) in Java, which is built with different
language constraints:

```java
class Main {
  public static void main(String[] args) {
    String name = "Danger";
    System.out.println("Hello, " + name);
  }
}
```

> > Note: if you find the naming of Java and JavaScript confusing (it is, they are two completely separate

Reviewer: ```suggestion
> Note: if you find the naming of Java and JavaScript confusing, it is (they are two completely separate
```

> programming languages, no link at all.) JavaScript was named that way because Java was looking to be really the
> next hot language (it did turn out that way for a decade or two, but now JavaScript is usually the first language

> > people have heard of.)

Reviewer: This could link out to a resources to learn more, maybe this one? https://medium.com/@benastontweet/lesson-1a-the-history-of-javascript-8c1ce3bffb17

Aside from having a lot more lines, the Java version comes with a lot of words that aren't necessarily about
telling the computer exactly what to do, e.g. `class Main {`, `public static void main(String[] args) {`, `}` and

> `}` again. It also has semi-colons at the end of some lines. Java is aimed at building different things, and these

Reviewer: ```suggestion
`}` again. It also has semi-colons at the end of some lines. Java is aimed at building different things from JavaScript, and these
```

extra bits of code make sense within the constraints of building a Java app.

To get to my main point though, there is one standout line I'd like us to compare:

```
// JavaScript
var name = "Danger"
// Java
String name = "Danger";
```

Both of these lines declare variables called `name` which contain the value `"Danger"`.

In JavaScript you use the abbreviation `var` to declare a variable. Meanwhile, in Java you need to say _what kind
of data_ the variable contains. In this case the variable contains a `String`. (A string is a programming term for
a collection of characters. They `"look like this"`. This [5m video](https://www.youtube.com/watch?v=czTWbdwbt7E)
is a good primer if you want to learn more.)

Both of these variables contain a string, but the difference is that in Java the variable can _only ever contain a
\_string_, because that's what we said when we created the variable. In JS the variable can change to be
_anything_, like a number, or a list of dates.

There are many values which can go into a variable, and `var` accepts all of them in JS. In Java though, we have to

> say upfront what can be inside that variable, which is only strings, so:

Reviewer 2: This paragraph is a bit redundant after the ones above changed. Could be changed to just 'To illustrate:'

```js
// Before in JS
var name = "Danger"
// Also OK
var name = 1
var name = false
var name = ["2018-02-03", "2019-01-12"]

// Before in Java
String name = "Danger";
// Not OK, the code wouldn't be accepted by Java
String name = 1;
String name = false

> String name = {"2018-02-03", "2019-01-12"};

Reviewer 2: I don't think `{"2018-02-03", "2019-01-12"}` is a valid java expression. Did you mean `new String[]{"2018-02-03", "2019-01-12"}`? Or just removing noise to help beginners?

Author: I threw it into a Java VM and it errored like I thought it would - guess I didn't know how to make it actually succeed the assignment - switching to the working version

```

These trade-offs make sense in the context for which these languages were built back in 1995. JavaScript was
originally designed to be a small programming language which handled simple interactions on websites. Java on the
other hand was built specifically to make big apps which could run on any computer. Their needs had different

> scales, so the language required different types of code.

Reviewer 2: Much clearer 👍

Java required programmers to be more explicit with the values of their variables because the programs they expected
people to build were more complex. While JavaScript opted for ease of reading, and flexibility because you would be

> able to see most of the code in one go.

Reviewer: > be able to see most of the code in one go

This ideas needs a little more to it. Java was a language built for enterprise scale and JavaScript was developed when web interactivity was very nascent; is it okay to just say that JavaScript apps had less code because they initially did less stuff than Java apps?

Author: agree, done that 👍

### What is TypeScript?

TypeScript is a programming language - it contains all of JavaScript, and then a bit more. Using our example above,
let's compare the scripts for "Hello, Danger" in JavaScript vs TypeScript:

```js
// JavaScript
var name = "Danger"
console.log("Hello, " + danger)

// TypeScript
var name = "Danger"
console.log("Hello, " + danger)

// Yep, you're not missing something, there's no difference
```

Due to TypeScript's aim to only _extend_ JavaScript, your normal JavaScript code should work fine with TypeScript.
The things TypeScript adds to JavaScript are intended to help you be more explicit about what kinds of data are
used in your code, a bit like Java.

```diff
- var name = "Danger"
+ var name: string = "Danger"
console.log("Hello, " + danger)
```

This extra `: string` allows the reader to be certain that `name` will only be a string. Annotating your variables

> also allows TypeScript to check this for you. This is very useful because keeping track of changes like the type of

Reviewer: ```suggestion
also allows TypeScript to check this for you. This is _very_ useful because keeping track of changes like the type of
```

value in a variable seems easy when it's one or two, but once it starts hitting the hundreds, that's a lot to keep

> track of.

Reviewer: ```suggestion
track of. Types help programmers be more confident about their code because types catch mistakes.
```

Simply speaking, we call these annotations "Types". Hence the name <i>Type</i>Script. The tag-line for TypeScript
is "JavaScript which scales" which is a statement that these extra type annotations allows you to work on bigger
projects. This is because you can verify up-front how correct your code is. This means you have less need to
understand how every change affects the rest of the program.

> 

Reviewer: This section could use a summarizing paragraph. It's not clear why TypeScript is appropriate _now_, while JavaScript not having types made sense in the 90's. Throughout this post, we've contrasted the two languages, which is a lot to grok, and we should reward the reader with an understanding of why the tradeoffs worked out the way they did.

Author: perfect, this was a killer point!

### Why does Artsy use TypeScript?

Artsy definitely isn't the size of Microsoft! Artsy is about 30 engineers, and Microsoft are about 60,000. However,
some of our problems are the same. Developers at Artsy build apps which are made up of thousands of files. Any
single change to one file, could affect any other file. A change to one individual file can affect the behaviour of
any number of other files, like throwing a pebble into a pond and causing ripples to spread out to the bank.

Typically, the need to ensure there are no bugs is less of a problem for people building websites. Websites are
easy to make changes to, because if you change the site - everyone gets the update instantly. However, we also

> build our iOS app with JavaScript, and a change to the app requires Apple to review the changes and for someone to

Reviewer: ```suggestion
build our iOS app with JavaScript, and a change to the app requires Apple to review the changes and for users to
```

download the new version from the App Store.

This means that the iOS team needs to have more checks that everything is OK before shipping the app to the world.
Using TypeScript gives our team the ability to feel good that the changes we have made are only the changes we
want.

TypeScript isn't the only programming language to tackle the problem of making JavaScript code safer, but it's the

> one with the biggest community, allows people to re-use their JavaScript knowledge and has really good tools to

Reviewer: More than getting to re-use knowledge, we can re-use it across multiple domains. Web developers can write iOS code, for example. 

There's also how TypeScript makes it easy to migrate codebases gradually over time, which is a huge benefit that might not occur to non-programmers.

Author: This one is tricky, I'm currently framing TS vs other transpiled languages which would have  all of the same runtime advantages as TS

Yeah, I've not been able to figure out the migration angle - but I've added a note

content that will appear on the page, CSS defines the visual style of the page, and JS defines the interactive
behaviours of the page.

We describe having these sets of skills as being a "front-end" developer. You have to understand those three
languages to present anything inside a web browser like Safari, Firefox or Chrome. So, given how popular the web
is, there is a massive demand for people who are good at using these three languages.

There is also the set of skills for the "back-end" developers, which are to create computer services that
communicate either to a web browser (by passing it HTML/CSS/JS) or to another service (by sending a raw data.) You
don't need to use HTML, CSS or JS to write this type of code, but it's usually an end-product of your work. We
mostly build our back-ends in Ruby or JavaScript at Artsy.

### What do Programming Languages do?

Programming languages are an interesting problem to solve. People read code many, many multiples of times more than
they write it - so developers create languages which are good at solving particular problems with a small amount of
code. Here's an example using JavaScript:

```js
var name = "Danger"

> console.log("Hello, " + danger)

Author: hah, great points!

> console.log("Hello, " + danger)

Reviewer 2: What delicious irony that TypeScript would have caught these errors 🤦‍♂️😑
